FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+
+            // step 5: update jobInfo entity's status to running, also update version and heartbeat
+            jobInfo.Status = JobStatus.Running;
+
+            // jobInfo's version is set when dequeue, if there are multi running jobs for this jobInfo, only the last one will keep alive
+            jobInfo.Version = DateTimeOffset.UtcNow.Ticks;
+            jobInfo.HeartbeatDateTime = DateTime.UtcNow;
+            jobInfo.HeartbeatTimeoutSec = heartbeatTimeoutSec;
+            var jobInfoEntity = jobInfo.ToTableEntity();
+
+            // step 6: update message pop receipt to job lock entity
+            jobLockEntity.JobMessagePopReceipt = message.Value.PopReceipt;
+
+            // step 7: transaction update jobInfo entity and job lock entity
+            IEnumerable<TableTransactionAction> transactionUpdateActions = new List<TableTransactionAction>
+            {
+                new (TableTransactionActionType.UpdateReplace, jobInfoEntity, jobInfoEntityResponse.Value.ETag),
+                new (TableTransactionActionType.UpdateReplace, jobLockEntity, jobLockEntity.ETag),
+            };
+
+            _ = await _azureJobInfoTableClient.SubmitTransactionAsync(transactionUpdateActions, cancellationToken);
+
+            _logger.LogInformation($"Dequeue job '{jobInfo.Id}'. Successfully ");
+            return jobInfo;
+        }
+
+        public async Task<JobInfo> GetJobByIdAsync(byte queueType, long jobId, bool returnDefinition, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to get job {jobId}.");
+
+            // step 1: get job reverse index entity
+            var jobReverseIndexEntity = await GetJobReverseIndexEntityByIdAsync(queueType, jobId, cancellationToken);
+
+            // step 2: get job info entity
+            var selectedProperties = returnDefinition ? null : SelectPropertiesExceptDefinition();
+
+            var jobInfoEntityResponse = await _azureJobInfoTableClient.GetEntityAsync<JobInfoEntity>(
+                jobReverseIndexEntity.JobInfoEntityPartitionKey,
+                jobReverseIndexEntity.JobInfoEntityRowKey,
+                selectedProperties,
+                cancellationToken);
+            var jobInfoEntity = jobInfoEntityResponse.Value;
+
+            // step 3: convert to job info.
+            var jobInfo = jobInfoEntity.ToJobInfo<TJobInfo>();
+
+            _logger.LogInformation($"Get job {jobId} successfully.");
+            return jobInfo;
+        }
+
+        public async Task<IEnumerable<JobInfo>> GetJobsByIdsAsync(byte queueType, long[] jobIds, bool returnDefinition, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to get jobs {string.Join(",", jobIds)}.");
+
+            var result = new List<JobInfo>();
+            foreach (var id in jobIds)
+            {
+                // get each job by id
+                result.Add(await GetJobByIdAsync(queueType, id, returnDefinition, cancellationToken));
```

Comment:
Maybe we can use `Task.WhenAll` here

Reply:
Thanks for the good suggestion, I will refine it.

Reply:
[Parallel.ForEachAsync](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreachasync?view=net-6.0) is used.